### PR TITLE
Aggregate cross-host data transfers made by jax.device_put

### PR DESCRIPTION
**Pull request by rao-ashish**
This PR modifies `_batched_device_put_impl` to batch the cross-host data transfers of multiple arrays. This enables us to take advantage of recent optimizations inside XLA for cross-host data transfers, particularly on GPUs ([XLA #33284](https://github.com/openxla/xla/pull/33284), [XLA #33910](https://github.com/openxla/xla/pull/33910), [XLA #34904](https://github.com/openxla/xla/pull/34904)). `CrossHostTransferTest.test_cross_host_transfer_batched` is lightly modified and enabled for the CUDA platform to test this change.

### CODE REVIEW

File: jax/_src/dispatch.py [hunk 1]
```diff
@@ -571,18 +595,31 @@ def _batched_device_put_impl(
...
-    # efficient batch transfer.
     # device_put handles `Format` via a different path, so just pass `None` as
     # the layout here.
     shard_arg_results = pxla.shard_args(dsa_shardings, [None] * len(dsa_xs),
                                         dsa_copy_semantics, dsa_xs)
     for i, shard_arg_result in zip(dsa_indices, shard_arg_results):
       assert isinstance(ys[i], _DeferredShardArg)
       ys[i] = ys[i].result_handler(shard_arg_result)
+  if dca_xs:
+    copy_array_results = xc.batched_copy_array_to_devices_with_sharding(
```
Comment (yashk2810) [inline jax/_src/dispatch.py:607]: Why doesn't the current path take care of this? pxla.shard_args should call into `batched_copy_array_to_devices_with_sharding`.

The fix should be above where host transfers is triggered to actually create a DeferredShardArg.

Comment (yashk2810) [reply jax/_src/dispatch.py:607]: I would rather build a bit called `cross_host_xfer` into the current DeferredShardArg and dispatch it like that instead of replicating this complicated logic.

**Review by yashk2810 — CHANGES_REQUESTED** [review]

**Comment by rao-ashish** [comment]
Thanks for reviewing this! I looked into the `pxla.shard_args` code, and I think we will need to handle cross-host transfers separately by directly calling `batched_copy_array_to_devices_with_sharding`.

`pxla.shard_args` for jax Arrays eventually calls `_shard_np_array` for each array, which breaks the batching of transfers. Further, this calls into jaxlib's `batched_device_put` method with a batch comprised of shards local to a single array. `batched_device_put` ultimately calls into `DevicePutWithSharding` inside `py_values.cc`, which also specializes in the cross-host case to call `ifrt_client->CopyArrays(...)` for one shard at a time. So, the most feasible way to implement batching would be to explicitly call `xc.batched_copy_array_to_devices_with_sharding` like the original code was doing.

However, I've removed the new dataclass and added a field inside `_DeferredShardArgs` to designate cross-host transfers as you suggested. I've also introduced a function `_populate_batched_results` to reduce code duplication between the two cases.

**Comment by yashk2810** [comment]
I think we can merge and simplify more. Let me patch this change in internally and try things out.

**Comment by rao-ashish** [comment]
Hi, just wanted to follow up to see if you were able to figure out a nicer way to batch transfers with device_put? Or should we go with a separate call to `batched_copy_array_to_devices_with_sharding` as done in this PR?

**Comment by yashk2810** [comment]
I didn't get time yet but let's revert back to the way you had it before. The current change is a bit worse than the one you had before.

I'll refactor it after you submit the CL.

**Comment by rao-ashish** [comment]
Thanks! I've just removed the second commit to go back to the first version.

**Comment by rao-ashish** [comment]
Just double checking if this looks good now / is ready to be approved?

**Review by yashk2810 — APPROVED** [review]
